src/common/overlay.c

```diff
@@ -126,18 +126,25 @@ GList *dt_overlay_get_used_in_imgs(const dt_imgid_t overlay_id,
   return res;
 }
 
-gboolean dt_overlay_used_by(const dt_imgid_t imgid, const dt_imgid_t overlay_id)
+gboolean dt_overlay_used_by(const dt_imgid_t imgid_intended_overlay, const dt_imgid_t imgid_target_image)
```

Comment:
I did not swap params, only gave them what I think are more meaningful names.
Here is the code calling this function (from `iop/overlay.c _drag_and_drop_received`):
```c
      dt_imgid_t *imgs = (dt_imgid_t *)gtk_selection_data_get_data(selection_data);

      const dt_imgid_t imgid = imgs[0];

      // check for cross-references, that is this imgid should not be using
      // the current image as overlay.

      if(dt_overlay_used_by(imgid, self->dev->image_storage.id))
      {
        dt_control_log
          (_("cannot use image %d as an overlay"
             " as it is using the current image as an overlay, directly or indirectly"),
           imgid);
      }
```
As you can see, the 1st param, **`imgid` above, is** that of **the dropped image, intended to be used as an overlay**; the 2nd, **`self->dev->image_storage.id`** in the calling code, belongs to the **the current image**.
In the function, they used to be named based on what fields in the `overlay` table they would be matched against; that is because we want to check if dropped image (the intended overlay) already uses the current image (the target) as an overlay, leading to a loop.

Reply:
I don't think so.

```c
      dt_imgid_t *imgs = (dt_imgid_t *)gtk_selection_data_get_data(selection_data);

      const dt_imgid_t imgid = imgs[0];
```
`imgid` is the ID of the dropped image, the overlay.
`self->dev->image_storage.id` is the ID of the current image.

The corresponding query in the original code is:
```c
gboolean dt_overlay_used_by(const dt_imgid_t imgid, const dt_imgid_t overlay_id)
{
  sqlite3_stmt *stmt;

  DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db),
                              "SELECT 1"
                              " FROM overlay"
                              " WHERE imgid = ?1" -- is set to imgid, the 1st param
                              "   AND overlay_id = ?2", -- is set to overlay_id, the 2nd param
                              -1, &stmt, NULL);
  DT_DEBUG_SQLITE3_BIND_INT(stmt, 1, imgid);
  DT_DEBUG_SQLITE3_BIND_INT(stmt, 2, overlay_id);
```
So, in the query, we want to check the following: is there already an association, where the `imgid` column is the 1st, `imgid` parameter, and the `overlay_id` column is the 2nd, `overlay_id` parameter? If yes, then `imgid` cannot be used as the **overlay** of the current image (identified by the `overlay_id` parameter), because we would have a circular dependency.
If we find something, we display the warning, saying we cannot use `imgid` as the **overlay**.
Note that `imgid` is clearly indicated as the ID of the image we want to use as the overlay. Utterly confusing, I think:
```c
if(dt_overlay_used_by(imgid, self->dev->image_storage.id))
      {
        dt_control_log
          (_("cannot use image %d as an overlay"
             " as it is using the current image as an overlay, directly or indirectly"),
           imgid);
      }
```

If we don't find anything, we record the new association:
```c
        p->imgid         = imgid;
        _clear_cache_entry(self, index);

        dt_overlay_record(self->dev->image_storage.id, p->imgid);
```
Where `dt_overlay_record` uses the current image ID as the 1st param, the imgid as the 2nd param, and the function is:
```c
void dt_overlay_record(const dt_imgid_t imgid, const dt_imgid_t overlay_id)
{
  sqlite3_stmt *stmt;
  DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db),
                              "INSERT INTO overlay (imgid, overlay_id) "
                              "VALUES (?1, ?2)",
                              -1, &stmt, NULL);

  DT_DEBUG_SQLITE3_BIND_INT(stmt, 1, imgid);
  DT_DEBUG_SQLITE3_BIND_INT(stmt, 2, overlay_id);
```
So the 1st param (from set from `self->dev->image_storage.id`) is the image ID to which we apply the overlay, the 2nd (set from `imgid`) is the ID of the image we set as the overlay.

Maybe it would be best if I renamed the variables in the calling function, to be consistent with the query params:
```c
      dt_imgid_t *imgs = (dt_imgid_t *)gtk_selection_data_get_data(selection_data);

      const dt_imgid_t imgid_intended_overlay = imgs[0];
      const dt_imgid_t imgid_target_image = self->dev->image_storage.id;

      // check for cross-references, that is this imgid_intended_overlay should not be using
      // the current image as overlay.

      if(dt_overlay_used_by(imgid_intended_overlay, imgid_target_image))
      {
        dt_control_log
          (_("cannot use image %d as an overlay"
             " as it is using the current image as an overlay, directly or indirectly"),
           imgid_intended_overlay);
      }
      else
      {
        // remove previous overlay if valid
        if(dt_is_valid_imgid(p->imgid))
          dt_overlay_remove(imgid_target_image, p->imgid);

        // and record the new one
        p->imgid         = imgid_intended_overlay;
        _clear_cache_entry(self, index);

        dt_overlay_record(imgid_target_image, imgid_intended_overlay);

        dt_image_full_path(imgid_intended_overlay, p->filename, sizeof(p->filename), NULL);
```

Reply:
Done, please check.